sdks/java/io/synthetic/src/main/java/org/apache/beam/sdk/io/synthetic/SyntheticStep.java

```diff
@@ -0,0 +1,203 @@
+      c.output(outputElement(key, val, hashCodeOfVal, i, random));
+    }
+    if (random.nextDouble() < fractionalPart) {
+      c.output(outputElement(key, val, hashCodeOfVal, i, random));
+    }
+  }
+
+  @StartBundle
+  public void startBundle() throws Exception {
+    if (options.perBundleDelay > 0) {
+      SyntheticUtils.delay(
+          Duration.millis(options.perBundleDelay),
+          options.cpuUtilizationInMixedDelay,
+          options.perBundleDelayType,
+          new Random());
+    }
+  }
+
+  /**
+   * Synthetic step options. These options are all JSON, see documentations of individual fields for
+   * details. {@code Options} uses jackson annotations which PipelineOptionsFactory can use to parse
+   * and construct an instance.
+   */
+  public static class Options extends SyntheticOptions {
+    /**
+     * Amplification/filtering ratio: the number of output records should be emitted on average for
+     * each input record.
+     */
+    @JsonProperty public double outputRecordsPerInputRecord;
+
+    /**
+     * If false, the DoFn generates a different distribution of KV pairs according to the parameters
+     * in {@link SyntheticOptions}, and input records are merely used as a “clock”; If true, the
+     * shape of the input distribution is preserved, and the DoFn only does sleeping and
+     * amplification/filtering.
+     */
+    @JsonProperty public boolean preservesInputKeyDistribution;
+
+    /** User-defined tag to annotate synthetic step name. */
+    @JsonProperty public String uniqueTag;
+
+    /**
+     * An upper limit on throughput across the worker for this step. In a streaming job, it is not
+     * easy to tightly control parallelism of a DoFn. It depends on various factors. As a result, it
+     * is much harder to control throughput preconfigured cpu or sleep delay. When max throughput is
+     * set, SyntheticStep delays beyond the configured delay (either cpu or sleep) in order to keep
+     * the overall throughput below the limit.
+     */
+    @JsonProperty public long maxWorkerThroughput = -1;
+
+    /**
+     * Number of milliseconds to delay for in each bundle. Cannot be enabled simultaneously with
+     * maxWorkerThroughput >= 0.
+     */
+    @JsonProperty public long perBundleDelay = 0;
+
+    /** Type of per bundle delay to use ("SLEEP", "CPU", "MIXED"). */
+    @JsonProperty public DelayType perBundleDelayType = DelayType.SLEEP;
+
+    /**
+     * If true, reports time spent sleeping as 'cumulativeThrottlingMicros' metric. This enables
+     * Dataflow to detect throttled stages, which would influence scaling decisions.
+     */
+    @JsonProperty public boolean reportThrottlingMicros;
+
+    @Override
+    public void validate() {
+      super.validate();
+      checkArgument(
+          outputRecordsPerInputRecord >= 0,
+          "outputRecordsPerInputRecord should be a non-negative number, but found %s.",
+          outputRecordsPerInputRecord);
+      if (maxWorkerThroughput >= 0) {
```

Comment:
If those both variables can't be used together, don't we have to check what happens if `perBundleDelay != 0` too?

Reply:
I've added a `checkArgument` call to make sure that it has a correct value. If `perBundleDelay > 0` and `maxWorkerThroughput > 0` (the only erroneous condition), then the error will be triggered.